src/random_ray/random_ray.cpp

```diff
@@ -0,0 +1,284 @@
+#include "openmc/random_ray/random_ray.h"
+#include "openmc/geometry.h"
+#include "openmc/message_passing.h"
+#include "openmc/mgxs_interface.h"
+#include "openmc/random_ray/source_region.h"
+#include "openmc/settings.h"
+#include "openmc/simulation.h"
+#include "openmc/source.h"
+
+namespace openmc {
+
+//==============================================================================
+// RandomRay implementation
+//==============================================================================
+
+RandomRay::RandomRay()
+{
+  angular_flux_.resize(data::mg.num_energy_groups_);
+  delta_psi_.resize(data::mg.num_energy_groups_);
+}
+
+RandomRay::RandomRay(uint64_t index_source) : RandomRay::RandomRay()
+{
+  initialize_ray(index_source);
+}
+
+// Transports ray until termination criteria are met
+uint64_t RandomRay::transport_history_based_single_ray()
+{
+  using namespace openmc;
+  while (true) {
```

Comment:
How about `while alive())` instead, removing that first conditional?

Reply:
I like it! I'll make the change.